src/authn/index.js

```diff
@@ -119,32 +150,64 @@ function setup(app) {
-      serializedUser.authzRoles = [...serializedUser.authzRoles];
+    // Deflate authzRoles and flags from Set to Array
+    for (const key of ["authzRoles", "flags"]) {
+      if (key in serializedUser) {
+        serializedUser[key] = [...serializedUser[key]];
+      }
     }
 
     return done(null, JSON.stringify(serializedUser));
   });
 
-  passport.deserializeUser((serializedUser, done) => {
+  passport.deserializeUser((req, serializedUser, done) => {
+    /* Mark the request as being authenticated with a session, which is
+     * strongly indicative of a browser client (as API clients use tokens).
+     */
+    req.context.authnWithSession = true;
+
     let user = JSON.parse(serializedUser); // eslint-disable-line prefer-const
 
-    // Inflate authzRoles from Array back into a Set
-    if ("authzRoles" in user) {
-      user.authzRoles = new Set(user.authzRoles);
+    // Inflate authzRoles and flags from Array back into a Set
+    for (const key of ["authzRoles", "flags"]) {
+      if (key in user) {
+        user[key] = new Set(user[key]);
+      }
     }
 
-    /* Update existing sessions that pre-date the presence of authzRoles.  When
-     * authzRoles is absent, "groups" is the unparsed set of Cognito groups.
-     */
-    if (!("authzRoles" in user)) {
-      const {groups, authzRoles} = parseCognitoGroups(user.groups || []);
-      utils.verbose(`Upgrading groups and authzRoles of user object for ${user.username}`);
+    const update = (originalCognitoGroups, reason) => {
+      const {groups, authzRoles, flags, cognitoGroups} = parseCognitoGroups(originalCognitoGroups);
+      utils.verbose(`Updating user object for ${user.username}: ${reason}`);
       user.groups = groups;
       user.authzRoles = authzRoles;
+      user.flags = flags;
+      user.cognitoGroups = cognitoGroups;
       user[RESAVE_TO_SESSION] = true;
-    }
+    };
+
+    /* Update existing sessions that pre-date the presence of authzRoles.  When
+     * authzRoles is absent, "groups" is the unparsed set of Cognito groups.
+     */
+    if (!("authzRoles" in user)) update(user.groups || [], "missing authzRoles");
+
+    /* Update existing sessions that pre-date the presence of flags.  When
+     * flags is absent, then authzRoles is the unparsed set of Cognito groups.
+     *
+     * In almost all cases, flags will be an empty set.  However, if a user is
+     * added to a flags group (or other internal group) in Cognito *and* then
+     * also establishes a new session in production before this new flags
+     * support (i.e. internal groups support) support is deployed to
+     * production, then that session's user object will be wrong and we'll fix
+     * it here.  Such a wrong user object would have, for example, a "!flags"
+     * Nextstrain Group with role "!flags/canary".  This is likely to only
+     * impact internal team members.
+     */
+    if (!("flags" in user)) update([...user.authzRoles], "missing flags");
```

Comment:
What's the reasoning behind having `user.groups || []` above and `[...user.authzRoles]` here?

Reply:
If you mean why the different syntax, the reason is the function takes an array and `user.groups` is an array already but `user.authzRoles` is a Set, so `[...user.authzRoles]` converts it to an array.

If you mean why the different properties, the reason is that these conditional branches act on different past versions/schemas of the session-persisted user object, so the property which contains the "original Cognito groups array" in each is different (as noted in the comments above each).